dot/telemetry/telemetry.go

```diff
@@ -96,29 +100,87 @@ type ConnectionData struct {
 		"implementation": data.SystemName, "msg": "system.connected", "name": data.NodeName, "network_id": data.NetworkID, "startup_time": data.StartTime,
 		"version": data.SystemVersion}
-	h.telemetryLogger = log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
-	h.telemetryLogger.Print()
-	h.sendTelemtry()
+	telemetryLogger := log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
+	telemetryLogger.Print()
 }
 
 // SendBlockImport sends block imported message to telemetry connection
 func (h *Handler) SendBlockImport(bestHash string, height *big.Int) {
+	h.Lock()
+	defer h.Unlock()
 	payload := log.Fields{"best": bestHash, "height": height.Int64(), "msg": "block.import", "origin": "NetworkInitialSync"}
-	h.telemetryLogger = log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
-	h.telemetryLogger.Print()
-	h.sendTelemtry()
+	telemetryLogger := log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
+	telemetryLogger.Print()
 }
 
-func (h *Handler) sendTelemtry() {
-	for _, c := range h.wsConn {
-		err := c.WriteMessage(websocket.TextMessage, h.buf.Bytes())
+// NetworkData struct to hold network data telemetry information
+type NetworkData struct {
+	peers   int
+	rateIn  float64
+	rateOut float64
+}
+
+// NewNetworkData creates networkData struct
+func NewNetworkData(peers int, rateIn, rateOut float64) *NetworkData {
+	return &NetworkData{
+		peers:   peers,
+		rateIn:  rateIn,
+		rateOut: rateOut,
+	}
+}
+
+// SendNetworkData send network data system.interval message to telemetry connection
+func (h *Handler) SendNetworkData(data *NetworkData) {
+	h.Lock()
+	defer h.Unlock()
+	payload := log.Fields{"bandwidth_download": data.rateIn, "bandwidth_upload": data.rateOut, "msg": "system.interval", "peers": data.peers}
+	telemetryLogger := log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
+	telemetryLogger.Print()
+}
+
+// BlockIntervalData struct to hold data for block system.interval message
+type BlockIntervalData struct {
+	BestHash           common.Hash
+	BestHeight         *big.Int
+	FinalizedHash      common.Hash
+	FinalizedHeight    *big.Int
+	TXCount            int
+	UsedStateCacheSize int
+}
+
+// SendBlockIntervalData send block data system interval information to telemetry connection
+func (h *Handler) SendBlockIntervalData(data *BlockIntervalData) {
+	h.Lock()
+	defer h.Unlock()
+	payload := log.Fields{"best": data.BestHash.String(), "finalized_hash": data.FinalizedHash.String(), // nolint
+		"finalized_height": data.FinalizedHeight, "height": data.BestHeight, "msg": "system.interval", "txcount": data.TXCount, // nolint
+		"used_state_cache_size": data.UsedStateCacheSize}
+	telemetryLogger := log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
+	telemetryLogger.Print()
+}
+
+func (h *Handler) sender() {
+	for {
+		h.RLock()
+		line, err := h.buf.ReadBytes(byte(10))
```

Comment:
Add a comment specifying that newline delimiter is used.

Reply:
added comment.